crates/analyzer-host/src/lsp_impl/active_initialized.rs

```diff
@@ -229,18 +229,58 @@ async fn on_set_trace(_: LspServerState, params: SetTraceParams, state: Arc<Asyn
 	Ok(())
 }
 
+async fn created_file(url: &Url, state: &Arc<AsyncRwLock<State>>) {
+	// incase it's a folder, get a list of files
+ 	let files = state.write().await.file_system.enumerate_folder(url.clone(), "*.p4".into()).await;
```

Comment:
Not sure if enumerate_folder() does what I want:

- Does it work if uri is a file?
- Does it recursively check folders inside it self?
- Does the uri sent as a argument get added to the return vector?

Reply:
The LSP based file system sends a request to the LSP Client. It's handled here:
https://github.com/p4lang/p4analyzer/blob/8056ffcfc3cbee609d84653549c37d439d000479/packages/p4-analyzer-vscode/src/extensions/LanguageClient.ts#L21

The URI is expected to be a URI for a workspace folder. Internally it uses a `workspace` object provided by VSCode. It will then return all the files found in that workspace folder (recursively), and as you surmized, it won't include the provided URI because it expects that to be a folder.

Looking generally at your code: Rather than using the `file_system` directly, we should be using the `WorkspaceManager` to retrieve the file and then coordinate with the `Analyzer`. The `File` we get back from the `WorkspaceManager` has a utility method called `is_open_in_ide()` that will return `true` if the file is currently active in VSCode. If it is, then we can ignore this file change event because VSCode is currently the source of truth for this file.

If the file is not open in VSCode then you can await on a call to `File::get_parsed_unit()`. If the file is not known to the workspace at this point, then it will be background processed and you'll receive the associated `FileId`. You can then call `Analyzer::update()` using the `FileId` you've just been given, and the text contents of the file by awaiting on `file_system.file_contents(file_uri)`.

The existing `on_text_document_did_change()` and `on_text_document_did_close()` handlers show how the `WorkspaceManager` and `Analyzer` can be coordinated through a `FileId`.

My understanding is that we won't receive these events for folders. When we dynamically register for this particula events we explicitly ask for `'.p4'` files to be watched:
https://github.com/p4lang/p4analyzer/blob/8056ffcfc3cbee609d84653549c37d439d000479/crates/analyzer-host/src/lsp_impl/initializing.rs#L61